File: data-structures/javascript/stack/stack.js
// Structured using the Object-Linked-To-Other-Objects (OLOO) principle
// describe by Kyle Simpson in his series 'You Don't Know JS
// 'this & Object Prototypes', Chaper 5

const stack = {
  top: null,
  push( node ) {
    if ( this.isEmpty() ) {
      this.top = node
    } else {
      node.next = this.top;
      this.top = node
    }
  },
  pop() {
    if ( this.isEmpty() ) {
      return 'Stack is empty, cannot pop!';
    } else {
      let item = this.top;
      this.top = this.top.next;
      return item;
    }
  },
  peek() {
    return this.top;
  },
  isEmpty() {
    return this.top === null ? true : false
  }
};


module.exports = stack;

File: data-structures/ruby/stack/stack.rb
class Stack

  attr_accessor :top

  def initialize
    @top = nil
  end

  def push( node )
    if self.isEmpty?
      self.top = node
    else
      node.next = self.top
      self.top = node
    end
  end

  def pop
    if self.isEmpty?
      raise StackError, 'Stack is empty, cannot pop!'
    else
      item = self.top
      self.top = self.top.next
      return item
    end
  end

  def peek
    return self.top
  end

  def isEmpty?
    return self.top == nil ? true : false
  end

  private

  class StackError < StandardError
  end

end
